### ISSUE TITLE
Can not MOVE or COPY file with HTTPS reverse proxy

### ISSUE THREAD

**Issue by phoenixsfly** [mar10/wsgidav#183]
I can not move, rename, copy any file.  the route is :
_client ==HTTPS=> nginx reserve proxy  ==HTTP=>wsgidav (cheroot)_

However, when I skip https reserve proxy
_client ==HTTP/HTTPS=>wsgidav (cheroot)_ 
everything is OK

Finally I found the problem. 

When I use https reserve proxy, the environ["HTTP_DESTINATION"] is "**https:**//abc.com/file" and the scheme is **https**

However the environ["wsgi.url_scheme"] is **http** , just because nginx access wsgidav using http

In this line of code  [#L942](https://github.com/mar10/wsgidav/blob/fa95132f8aee737533599868b2284f1a04380857/wsgidav/request_server.py#L942)
```
        if destScheme and destScheme.lower() != environ["wsgi.url_scheme"].lower():
            self._fail(
                HTTP_BAD_GATEWAY, "Source and destination must have the same scheme."
            )
```
it will check destination scheme and wsgi_url scheme and there is no doubt that it will fail.

I just simplify delete this line code and everying be OK.

So，this line code should be deleted ？

**Comment by mar10** [comment]
Hello,

this seems to be a known issue, e.g.:
  - http://silmor.de/proxysvn.phpx
  - https://forum.nginx.org/read.php?2,276838

I would not simply remove the check, since it is [compliant with the spec](https://tools.ietf.org/html/rfc4918#section-10.3):

> If the Destination value is an absolute-URI (Section 4.3 of
   [RFC3986]), it may name a different server (**or different port or
   scheme**).  If the source server cannot attempt a copy to the remote
   server, it MUST fail the request.
 
I can think of a few options:

1. Configure the reverse proxy to rewrite the `Destination` header's protocol from 'https:' to 'http:'
2. Modify WsgiDAV to also check for the `X-Forwarded-Proto` header and accept the request if this matches
3. Make the check (complete or the 2. variation) configurable

Did you try to configure Nginx ?
Is the `X-Forwarded-Proto` header passed to the request and what is its value?

**Comment by phoenixsfly** [comment]
Changing Destination header is useful. great!
In Nginx conf file, I set
`proxy_set_header X-Forwarded-Proto $scheme;`
so it would be https

**Comment by mar10** [comment]
Thanks for sharing. (I added a hint in the error message.)

**Comment by gabriele-v** [comment]
Hi @mar10
any chance you will add support for X-Forwarded-Proto ?

Now I'm rewriting destination on Nginx, but it's not the optimal solution IMHO..

**Comment by mar10** [comment]
@gabriele-v I currently don't have this scenario at hand, so no easy way to test.
But I am open to a PR, if you want to implement and confirm it works.

**Comment by gabriele-v** [comment]
I could understand that it's difficult to test, but that's not a reason to close the issue IMHO, because problem it's still there..
Regarding a PR, unfortunately (even if I could technically try to add some changes) it will be difficult for me to test because I'm using WSGIDAV integrated into Seafile.

**Comment by mdemoss** [comment]
I configured my nginx reverse proxy like this and it works:

```nginx
        location /drive/ { # wsgidav
                proxy_pass http://localhost:8922;
                proxy_http_version 1.1;
                proxy_set_header HOST $host;
                #proxy_set_header Connection "upgrade";
                proxy_set_header X-NginX-Proxy true;
                proxy_buffering off;
                client_max_body_size 0;
                proxy_request_buffering off;
                set $dest $http_destination;
                if ($http_destination ~ "^https://mycooldomain.example.com/drive/(.+)") {
                        set $dest /drive/$1;
                }
                proxy_set_header DESTINATION $dest;
        }
```

**Comment by gabriele-v** [comment]
Yes I've done the same on deatination (with a map rule) but IMHO the cleanest way should be X-Forwarded-Proto without other tricks inside Nginx, once protocol it's forwarded by proxy, than should be application to manage it.

**Comment by mar10** [comment]
It is not about the fix (which is probably simple: see #187), but it cannot be merged until someone steps up to test and confirm.

**Comment by Lobstros** [comment]
If it helps, I have replaced my copy of `wsgidav_app.py` with the fixed version from #187, and it works fine (running as part of Seafile 7.1.5)

**Comment by mar10** [comment]
Thanks @Lobstros !
